lib/jsdom/living/nodes/DOMTokenList-impl.js

```diff
@@ -0,0 +1,197 @@
+      return;
+    }
+
+    const val = getAttributeValue(this._element, this._attributeLocalName);
+    if (val === null) {
+      this._tokenSet.length = 0;
+    } else {
+      this._tokenSet = [...orderedSet.parse(val)];
+    }
+
+    this._dirty = false;
+  }
+
+  get tokenSet() {
+    this._syncWithElement();
+    return this._tokenSet;
+  }
+
+  _validationSteps(token) {
+    if (!this._supportedTokens) {
+      throw new TypeError(`${this._attributeLocalName} attribute has no supported tokens`);
+    }
+    const lowerToken = token.replace(/[A-Z]/g, l => l.toLowerCase());
+    return this._supportedTokens.has(lowerToken);
+  }
+
+  _updateSteps() {
+    if (!hasAttributeByName(this._element, this._attributeLocalName) && this._tokenSet.length === 0) {
+      return;
+    }
+    setAttributeValue(this._element, this._attributeLocalName, orderedSet.serialize(this._tokenSet));
+  }
+
+  _serializeSteps() {
+    return getAttributeValue(this._element, this._attributeLocalName);
+  }
+
+  get length() {
+    this._syncWithElement();
+    return this._tokenSet.length;
+  }
+
+  get [idlUtils.supportedPropertyIndices]() {
+    this._syncWithElement();
+    return this._tokenSet.keys();
+  }
+
+  item(index) {
+    this._syncWithElement();
+    if (index >= this._tokenSet.length) {
+      return null;
+    }
+    return this._tokenSet[index];
+  }
+
+  contains(token) {
+    this._syncWithElement();
+    return this._tokenSet.includes(token);
+  }
+
+  add(...tokens) {
+    for (const token of tokens) {
+      validateTokens(token);
+    }
+    this._syncWithElement();
+    for (const token of tokens) {
+      if (!this._tokenSet.includes(token)) {
+        this._tokenSet.push(token);
+      }
+    }
+    this._updateSteps();
+  }
+
+  remove(...tokens) {
+    for (const token of tokens) {
+      validateTokens(token);
+    }
+    this._syncWithElement();
+    if (tokens.length === 0 || this._tokenSet.length === 0) {
+      this._updateSteps();
+      return;
+    }
+    tokens = new Set(tokens);
+    for (let i = 0; i < this._tokenSet.length && tokens.size > 0;) {
+      if (tokens.has(this._tokenSet[i])) {
+        tokens.delete(this._tokenSet[i]);
+        this._tokenSet.splice(i, 1);
+      } else {
+        i++;
+      }
+    }
+    this._updateSteps();
+  }
+
+  toggle(token, force = undefined) {
+    validateTokens(token);
+    this._syncWithElement();
+    const idx = this._tokenSet.indexOf(token);
+    if (idx >= 0) {
+      if (force === undefined || force === false) {
+        this._tokenSet.splice(idx, 1);
+        this._updateSteps();
+        return false;
+      }
+      return true;
+    }
+    if (force === undefined || force === true) {
+      this._tokenSet.push(token);
+      this._updateSteps();
+      return true;
+    }
+    return false;
+  }
+
+  replace(token, newToken) {
+    validateTokens(token, newToken);
+    this._syncWithElement();
+
+    let seen = false;
+
+    for (let i = 0; i < this._tokenSet.length;) {
```

Comment:
This kind of complexity makes me wonder if a separate file/class for Infra's ordered set might be a good idea.